dali/kernels/signal/moving_mean_square_gpu.cu

```diff
@@ -0,0 +1,280 @@
+    // Out of bounds values are assumed to be 0.
+    for (int pos = threadIdx.x; pos < pow2; pos += blockDim.x) {
+      acc_t<In> value(0);
+      auto extended_blk_ptr = extended_blk_start + pos;
+      if (extended_blk_ptr >= input && extended_blk_ptr < extended_blk_end) {
+        value = *extended_blk_ptr;
+      }
+      temp[shm_pos(pos)] = pre(value);
+    }
+
+    // // Step 2: Calculate prefix sum of the extended block, in place
+    // // (note: __syncthreads already happens inside)
+    PrefixSumSharedMem(temp, pow2, shm_pos);
+
+    // Step 3: Compute the output, the sum in window, by subtracting two values of the prefix sum
+    // and adding the input value at the current position.
+    for (int pos = threadIdx.x; pos < logical_block_sz; pos += blockDim.x) {
+      acc_t<In> out_val = pre(logical_block_in_ptr[pos]) + temp[shm_pos(window + pos)] -
+                          temp[shm_pos(pos)];
+      logical_block_out_ptr[pos] = ConvertSat<Out>(post(out_val));
+    }
+  }
+}
+
+template <typename InputType>
+KernelRequirements MovingMeanSquareGpu<InputType>::Setup(KernelContext &ctx,
+                                                         const InListGPU<InputType, 1> &in) {
+  KernelRequirements req;
+  req.output_shapes = {in.shape};
+  return req;
+}
+
+
+template <typename InputType>
+void MovingMeanSquareGpu<InputType>::Run(KernelContext &ctx, const OutListGPU<float, 1> &out,
+                                         const InListGPU<InputType, 1> &in,
+                                         const MovingMeanSquareArgs &args) {
+  int nsamples = in.shape.num_samples();
+  auto *sample_descs_cpu = ctx.scratchpad->AllocatePinned<SampleDesc<float, InputType>>(nsamples);
+
+  int64_t max_len;
+  for (int i = 0; i < nsamples; i++) {
+    auto &sample = sample_descs_cpu[i];
+    sample.out = out[i].data;
+    sample.in = in[i].data;
+    sample.len = in[i].shape.num_elements();
+    max_len = std::max(sample.len, max_len);
+  }
+  auto *sample_descs_gpu =
+      ctx.scratchpad->ToGPU(ctx.gpu.stream, make_span(sample_descs_cpu, nsamples));
+
+  conflict_free_pos shm_pos;
+
+  int window_len = args.window_size;
+
+  int max_shm_bytes = GetSharedMemPerBlock();
+  int max_shm_elems = max_shm_bytes / sizeof(acc_t<InputType>);
+
+  // Get a power of two that doesn't exceed the desired maximum shared memory size
+  int pow2 = prev_pow2(max_shm_elems * SHARED_MEMORY_BANKS / (SHARED_MEMORY_BANKS + 1));
+
+  // If reset interval is given, selects the logical block so that is close to it
+  // effectively clearing accummulation error every `reset_interval` samples.
+  if (needs_reset<InputType>::value && args.reset_interval > 0 && args.reset_interval < pow2) {
+    pow2 = prev_pow2(args.reset_interval);
+  }
+
+  int shm_sz = shm_pos(pow2) * sizeof(acc_t<InputType>);
+  int logical_block = pow2 - window_len;
+  // Note: logical_block==1 is very wasteful, but better than failing
+  assert(logical_block > 0);
```

Comment:
Should it be assert or DALI_ENFORCE?

Reply:
Assert should do. If the logic is correct this should not happen.